packages/plugin-ext/src/main/browser/tabs/tabs-main.ts

```diff
@@ -15,28 +15,301 @@
+    }
+
+    protected generateTabId(tabTitle: Title<Widget>, groupId: number): string {
+        return `${groupId}~${tabTitle.owner.id}`;
+    }
+
+    protected createTabGroupDto(tabBar: TabBar<Widget>): TabGroupDto {
+        const oldDto = this.tabGroupModel.get(tabBar);
+        const groupId = oldDto ? oldDto.groupId : this.groupIdCounter++;
+        const tabs: TabDto[] = tabBar.titles.map(title => {
+            const tabDto = this.createTabDto(title, groupId);
+            return tabDto;
+        });
+        const viewColumn = 1;
+        return {
+            groupId, tabs, isActive: false, viewColumn
+        };
+    }
+
+    protected attachListenersToTabBar(tabBar: TabBar<Widget> | undefined): void {
+        if (!tabBar) {
+            return;
+        }
+        tabBar.titles.forEach(title => {
+            this.connectToSignal(this.disposableTabBarListeners, title.changed, this.onTabTitleChanged);
+        });
+
+        this.connectToSignal(this.disposableTabBarListeners, tabBar.tabMoved, this.onTabMoved);
+        this.connectToSignal(this.disposableTabBarListeners, tabBar.disposed, this.onTabGroupClosed);
+    }
+
+    protected evaluateTabDtoInput(widget: Widget): AnyInputDto {
+        if (widget instanceof EditorPreviewWidget) {
+            if (widget.editor instanceof MonacoDiffEditor) {
+                return {
+                    kind: TabInputKind.TextDiffInput,
+                    original: toUriComponents(widget.editor.originalModel.uri),
+                    modified: toUriComponents(widget.editor.modifiedModel.uri)
+                };
+            } else {
+                return {
+                    kind: TabInputKind.TextInput,
+                    uri: toUriComponents(widget.editor.uri.toString())
+                };
+            }
+            // TODO notebook support when implemented
+        } else if (widget instanceof ViewContainer) {
+            return {
+                kind: TabInputKind.WebviewEditorInput,
+                viewType: widget.id
+            };
+        } else if (widget instanceof TerminalWidget) {
+            return {
+                kind: TabInputKind.TerminalEditorInput
+            };
+        }
+
+        return { kind: TabInputKind.UnknownInput };
+    }
+
+    protected connectToSignal<T>(disposableList: Disposable[], signal: { connect(listener: T, context: unknown): void, disconnect(listener: T): void }, listener: T): void {
+        signal.connect(listener, this);
+        disposableList.push(Disposable.create(() => signal.disconnect(listener)));
+    }
+
+    protected tabDtosEqual(a: TabDto, b: TabDto): boolean {
+        return a.isActive === b.isActive &&
+            a.isDirty === b.isDirty &&
+            a.isPinned === b.isPinned &&
+            a.isPreview === b.isPreview &&
+            a.id === b.id;
+    }
+
+    protected getOrRebuildModel<T, R>(map: Map<T, R>, key: T): R {
+        // something broke so we rebuild the model
+        let item = map.get(key);
```

Comment:
Keeping the model correct is really finicky. Especially since we don't have nice central point to get the events from. while testing i had some rare occurences i could not really reproduce where it seemed that the timing of the events was of. No Idea how that could happen. Maybe it was just my error. Though it seems vscode has a similiar issiue since they are doing the same thing. Im not sure. Though this helps in case there still are timing issues in the code to keep the model consistent.